### ISSUE TITLE
pyright fails to recognize key existence in `TypedDict`s within class methods

### ISSUE THREAD

**Issue by Vulcagon** [microsoft/pyright#7589]
When checking if a key exists in a TypedDict using an if-statement, pyright still complains that this key might not exist in the TypedDict if the key is not required.

```python
from typing import NotRequired, TypedDict


class MyTypedDict(TypedDict):
    key: NotRequired[str]


class MyClass:
    def __init__(self) -> None:
        self.my_dict: MyTypedDict = {}

    def foo(self) -> None:
        key = 'key'

        if key in self.my_dict:
            value = self.my_dict[key]  # Error
        else:
            value = ...
```

> Could not access item in TypedDict
>   "key" is not a required key in "MyTypedDict", so access may result in runtime exception
> Pylance([reportTypedDictNotRequiredAccess](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportTypedDictNotRequiredAccess))

This seems to only occur in class methods

```python
from typing import NotRequired, TypedDict


class MyTypedDict(TypedDict):
    key: NotRequired[str]


def main() -> None:
    my_dict: MyTypedDict = {}
    key = 'key'

    if key in my_dict:
        value = my_dict[key]  # No Error
    else:
        value = ...


if __name__ == '__main__':
    main()
```

For some reason when I check if any string is in `my_dict` the error doesn't show up in both cases even though it should

```python
key = 'key'

if 'some string' in my_dict:
    value = my_dict[key]  # No Error
else:
    value = ...
```

But it shows up when checking if another variable is in my_dict in both cases

```python
key = 'key'
other_key = 'other key'

if other_key in my_dict:
    value = my_dict[key]  # Error
else:
    value = ...
```

This bug also occurs when creating a TypedDict like as follows

```python
from typing import TypedDict

MyTypedDict = TypedDict('MyTypedDict', {'key': str}, total=False)
```

I'm using pyright as VS Code extension and the error occurs in every type checking mode (besides "off" of course)

Python version: 3.12.2
VS Code version: 1.87.2 (Windows 11)
Python extension version: 2024.2.1
Pylance extension version: 2024.3.2

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by SarthakNikhal** [comment]
@erictraut Can I work on this issue?

**Comment by erictraut** [comment]
It is already fixed.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.357, which I just published.
